File: GitHubAuthentication/GitHubAuthenticationProvider.cs
using Microsoft.Kiota.Abstractions;
using Microsoft.Kiota.Abstractions.Authentication;

namespace GithubTodoDemo.GitHubAuthentication;

public class GitHubAuthenticationProvider : IAuthenticationProvider
{
	public GitHubAuthenticationProvider(string clientId, string scope, IEnumerable<string> validHosts)
	{
		if (string.IsNullOrEmpty(clientId))
			throw new ArgumentNullException(nameof(clientId));
		if (string.IsNullOrEmpty(scope))
			throw new ArgumentNullException(nameof(scope));
		if (validHosts == null)
			throw new ArgumentNullException(nameof(validHosts));

		AccessTokenProvider = new GitHubAccessTokenProvider {
			ClientId = clientId,
			Scope = scope,
			AllowedHostsValidator = new AllowedHostsValidator(validHosts)
		};
	}
	public IAccessTokenProvider AccessTokenProvider {get; private set;}
    private const string AuthorizationHeaderKey = "Authorization";
    private const string ClaimsKey = "claims";
	public async Task AuthenticateRequestAsync(RequestInformation request, Dictionary<string, object> additionalAuthenticationContext = null, CancellationToken cancellationToken = default) {
		if(request == null) throw new ArgumentNullException(nameof(request));
		request.Headers.Add("User-Agent", "GithubTodoDemo/1.0.0");
        if(additionalAuthenticationContext != null &&
            additionalAuthenticationContext.ContainsKey(ClaimsKey) &&
            request.Headers.ContainsKey(AuthorizationHeaderKey))
            request.Headers.Remove(AuthorizationHeaderKey);

        if(!request.Headers.ContainsKey(AuthorizationHeaderKey))
        {
            var token = await AccessTokenProvider.GetAuthorizationTokenAsync(request.URI, additionalAuthenticationContext, cancellationToken);
            if(!string.IsNullOrEmpty(token))
                request.Headers.Add(AuthorizationHeaderKey, $"Bearer {token}");
        }
	}
}

File: typescript/gitHubAuthentication/gitHubAuthenticationProvider.ts
import { AllowedHostsValidator, AuthenticationProvider, RequestInformation } from '@microsoft/kiota-abstractions';
import { GitHubAccessTokenProvider } from './gitHubAccessTokenProvider.js';

export class GitHubAuthenticationProvider implements AuthenticationProvider {
	private readonly accessTokenProvider: GitHubAccessTokenProvider;
	/**
	 *
	 */
	public constructor(clientId: string, scope: string, validHosts: string[] = ["api.github.com"]) {
		this.accessTokenProvider = new GitHubAccessTokenProvider(clientId, scope, new AllowedHostsValidator(new Set(validHosts)));
	}
	private static readonly authorizationHeaderKey = "Authorization";
    private static readonly claimsKey = "claims";
	public authenticateRequest: (request: RequestInformation, additionalAuthenticationContext?: Record<string, unknown>) => Promise<void> = async (request, additionalAuthenticationContext) => {
		if(!request) {
			throw new Error("request cannot be undefined");
		}
		request.headers.add("User-Agent", "GithubTodoDemo/1.0.0");
		if(additionalAuthenticationContext?.[GitHubAuthenticationProvider.claimsKey] &&
            request.headers.has(GitHubAuthenticationProvider.authorizationHeaderKey)) {
            request.headers.delete(GitHubAuthenticationProvider.authorizationHeaderKey);
		}
		if (!request.headers.has(GitHubAuthenticationProvider.authorizationHeaderKey)) {
			const token = await this.accessTokenProvider.getAuthorizationToken(request.URL, additionalAuthenticationContext);
			request.headers.add(GitHubAuthenticationProvider.authorizationHeaderKey, `Bearer ${token}`);
		}
	}

}